// socialite/lib/config.js
module.exports = {
      facebook: {
        root: "https://graph.facebook.com/",
        dialog: {
          auth: "https://www.facebook.com/dialog/oauth?",
          mobile: "https://m.facebook.com/dialog/oauth?",
          access: "https://graph.facebook.com/oauth/access_token?"
        },
        oauth: {
          version: '2'
        }
      },
      twitter: {
        root: "https://api.twitter.com/1.1/",
        oauth: {
          version: "1.0a",
          signature: "HMAC-SHA1"
        },
        dialog: {
          auth: "https://api.twitter.com/oauth/authenticate?",
          request: "https://twitter.com/oauth/request_token?",
          access: "https://twitter.com/oauth/access_token?"
        },
        endpoint: {
          stream: {
            public: "https://stream.twitter.com/1.1/",
            user: "https://userstream.twitter.com/1.1/",
            site: "https://sitestream.twitter.com/1.1/"
          }
        }
      },
      foursquare: {
        root: "https://api.foursquare.com/v2/",
        dialog: {
          auth: "https://foursquare.com/oauth2/authenticate?",
          access: "https://foursquare.com/oauth2/access_token?"
        },
        oauth: {
          version: '2'
        }
      },
      google: {
        root: 'https://maps.googleapis.com/',
        dialog: {
          auth: "https://accounts.google.com/o/oauth2/auth?",
          access: "https://accounts.google.com/o/oauth2/token?",
          revoke: "https://accounts.google.com/o/oauth2/revoke?"
        },
        oauth: {
          version: '2'
        }
      },
      bitly: {
        root: "https://api-ssl.bitly.com/v3/",
        dialog: {
          auth: "https://bitly.com/oauth/authorize?"
        },
        oauth: {
          version: '2'
        }
      },
      yelp: {
        root: "http://api.yelp.com/v2/",
        dialog: {
          auth: null,
          request: null,
          access: null
        },
        oauth: {
          version: "1.0",
          signature: "HMAC-SHA1"
        }
      },
      instagram: {
        root: "https://api.instagram.com/v1/",
        dialog: {
          auth: "https://api.instagram.com/oauth/authorize/?"
        },
        oauth: {
          version: '2'
        }
      },
      custom: {
        root: "",
        dialog: {
          auth: "",
          request: "",
          access: "",
          mobile: ""
        }
      },
  };
